### PR TITLE
Finish the new patient lookup flow

### PR DESCRIPTION
**Pull request by saket**

### CODE REVIEW

File: app/src/main/java/org/simple/clinic/search/PatientSearchScreen.kt [hunk 1]
```diff
@@ -146,4 +153,39 @@ class PatientSearchScreen(context: Context, attrs: AttributeSet) : RelativeLayou
...
+
+  fun setEmptyFullNameErrorVisible(visible: Boolean) {
+    fullNameInputLayout.error = if (visible) {
+      resources.getString(R.string.patientsearch_error_empty_fullname)
+    } else {
+      null
+    }
+  }
+
+  fun setEmptyDateOfBirthAndAgeErrorVisible(visible: Boolean) {
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/search/PatientSearchScreen.kt]: I feel there's a lot of common functionality between all the date of birth and age error functions. Can we extract this common stuff out?

File: app/src/main/java/org/simple/clinic/search/PatientSearchScreenController.kt [hunk 2]
```diff
@@ -20,15 +27,22 @@ private typealias Ui = PatientSearchScreen
...
+    private val dobValidator: DateOfBirthFormatValidator
 ) : ObservableTransformer<UiEvent, UiChange> {
 
   override fun apply(events: Observable<UiEvent>): ObservableSource<UiChange> {
-    val replayedEvents = events.compose(ReportAnalyticsEvents()).replay().refCount()
+    val replayedEvents = events
+        .compose(ReportAnalyticsEvents())
+        .replay()
+        .refCount()
+        .compose(validateQuery())
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/search/PatientSearchScreenController.kt]: We have reporting for all validation errors in `Patient Entry`. Should we be reporting these errors too?

File: app/src/main/java/org/simple/clinic/search/PatientSearchScreenController.kt [hunk 3]
```diff
@@ -76,21 +90,117 @@ class PatientSearchScreenController @Inject constructor(
...
+      val dateOfBirthChanges = events
+          .ofType<SearchQueryDateOfBirthChanged>()
+          .map { it.dateOfBirth.trim() }
+
+      val validationErrors = events.ofType<SearchClicked>()
+          .withLatestFrom(nameChanges, ageChanges, dateOfBirthChanges) { _, name, age, dob -> Triple(name, age, dob) }
+          .map { (name, age, dateOfBirth) ->
+            val errors = mutableListOf<PatientSearchValidationError>()
+
+            if (dateOfBirth.isNullOrBlank() && age.isNullOrBlank()) {
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/search/PatientSearchScreenController.kt]: I feel a `when` clause here might be better than a if-else-if clause. What do you think?

File: app/src/main/java/org/simple/clinic/search/PatientSearchScreenController.kt [hunk 4]
```diff
@@ -76,21 +90,117 @@ class PatientSearchScreenController @Inject constructor(
...
+              errors += BOTH_DATEOFBIRTH_AND_AGE_ABSENT
+
+            } else if (dateOfBirth?.isNotBlank() == true && age?.isNotBlank() == true) {
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/search/PatientSearchScreenController.kt]: What do you think about `dateOfBirth.isNullOrBlank().not()` vs `dateOfBirth?.isNotBlank() == true`?

File: app/src/main/java/org/simple/clinic/search/PatientSearchScreenController.kt [hunk 5]
```diff
@@ -76,21 +90,117 @@ class PatientSearchScreenController @Inject constructor(
...
+        .flatMapIterable { it.validationErrors }
+        .doOnNext { Analytics.reportInputValidationError(it.analyticsName) }
+        .map {
+          { ui: Ui ->
+            when (it) {
+              FULL_NAME_EMPTY -> ui.setEmptyFullNameErrorVisible(true)
+              BOTH_DATEOFBIRTH_AND_AGE_ABSENT -> ui.setEmptyDateOfBirthAndAgeErrorVisible(true)
+              INVALID_DATE_OF_BIRTH -> ui.setInvalidDateOfBirthErrorVisible(true)
+              DATE_OF_BIRTH_IN_FUTURE -> ui.setDateOfBirthIsInFutureErrorVisible(true)
+
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/search/PatientSearchScreenController.kt]: Unnecessary line?

File: app/src/main/java/org/simple/clinic/widgets/PrimarySolidButton.kt [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+/**
+ * Accepting better names.
+ *
+ * TODO: Draw bottom border.
+ * TODO: Draw touch feedback.
+ */
+class PrimarySolidButton(context: Context, attrs: AttributeSet?) : AppCompatButton(context, attrs) {
+
+  private val dpToPx = { dp: Int ->
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/widgets/PrimarySolidButton.kt]: Shouldn't this be a function? I don't see any benefit in declaring it as a lambda?

File: app/src/main/java/org/simple/clinic/widgets/PrimarySolidButton.kt [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    val resources = context.resources
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/widgets/PrimarySolidButton.kt]: FYI: We don't need to get the context for this. We can use `Resources.getSystem().displayMetrics`.

File: app/src/main/java/org/simple/clinic/search/PatientSearchScreenController.kt (hunk 2 above)
Comment (saket) [reply app/src/main/java/org/simple/clinic/search/PatientSearchScreenController.kt]: This is already being done inside `showValidationErrors()`

File: app/src/main/java/org/simple/clinic/search/PatientSearchScreenController.kt (hunk 4 above)
Comment (saket) [reply app/src/main/java/org/simple/clinic/search/PatientSearchScreenController.kt]: Looks much better. I'll change this.

File: router/src/main/java/org/simple/clinic/router/screen/BaseViewGroupKeyChanger.kt [hunk 8]
```diff
@@ -67,9 +67,9 @@ abstract class BaseViewGroupKeyChanger<T : Any> : KeyChanger {
           incomingView,
           direction,
           onCompleteListener = {
-            if (outgoingView != null && outgoingState != null) {
-              outgoingState.save(outgoingView)
-              removeOutgoingView(outgoingState, outgoingView)
+            if (outgoingView != null) {
```
[the remark this answers is not in the document]
Comment (saket) [reply router/src/main/java/org/simple/clinic/router/screen/BaseViewGroupKeyChanger.kt]: Sure

File: app/src/main/java/org/simple/clinic/widgets/PrimarySolidButton.kt (hunk 7 above)
Comment (saket) [reply app/src/main/java/org/simple/clinic/widgets/PrimarySolidButton.kt]: Oh yeaa how did I forget that

**Comment by saket** [comment]
@vinaysshenoy I think the validation errors are supposed to replace the hint instead of appearing together. At least, this is what we've done in all registration screens. 

pin

**Comment by saket** [comment]
@vinaysshenoy the PIN should be cleared if any validation error is shown

File: app/src/main/java/org/simple/clinic/search/results/PatientSearchResultsController.kt [hunk 9]
```diff
@@ -37,28 +39,17 @@ class PatientSearchResultsController @Inject constructor(
...
     return events.ofType<PatientSearchResultsScreenCreated>()
         .map { it.key }
-        .map { (_, age, dob) ->
-          val computedAge = when {
-            age.isNotBlank() -> age.trim().toInt()
-            else -> calculateAgeFromDateOfBirth(dob)
-          }
-          computedAge
-        }
+        .map { (_, age, dob) -> ageOrAgeFromDateOfBirth(age, dob) }
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/search/results/PatientSearchResultsController.kt]: This method naming seems weird. Might I suggest `coerceAgeFrom()` or `obtainAgeFrom()` instead?

File: app/src/main/res/layout/screen_patient_search_results.xml [hunk 10]
```diff
@@ -17,6 +17,47 @@
...
+      android:id="@+id/patientsearchresults_empty_state_text"
+      android:layout_width="wrap_content"
+      android:layout_height="wrap_content"
+      android:layout_centerInParent="true"
+      android:layout_marginEnd="@dimen/spacing_24"
+      android:layout_marginStart="@dimen/spacing_24"
+      android:text="@string/patientsearchresults_empty_state"
+      android:textAppearance="@style/Clinic.V2.TextAppearance.H6.Grey2" />
+
+    <View
```
Comment (vinaysshenoy) [inline app/src/main/res/layout/screen_patient_search_results.xml]: Isn't there a widget named `Space` that's generally meant for this purpose?

Comment (saket) [reply app/src/main/res/layout/screen_patient_search_results.xml]: This is not empty state. It's the tail of the arrow. 😄

File: app/src/main/java/org/simple/clinic/search/PatientSearchScreen.kt (hunk 1 above)
Comment (saket) [reply app/src/main/java/org/simple/clinic/search/PatientSearchScreen.kt]: Done

**Pull request merged**